transformer_engine/pytorch/ops/op.py

```diff
@@ -0,0 +1,428 @@
+            meta.scale_inv = torch.ones(num_scales, dtype=dtype, device=device)
+            meta.amax_history = torch.zeros(
+                (recipe.amax_history_len, num_scales),
+                dtype=dtype,
+                device=device,
+            )
+            return {
+                key: meta,
+                "recipe": recipe,
+                "fp8_group": None,
+            }
+
+        # Construct FP8 metadata for all tensor types
+        return dict(
+            input=_make_meta(self.num_fp8_scales("input"), True),
+            param=_make_meta(self.num_fp8_scales("param"), True),
+            grad_output=_make_meta(self.num_fp8_scales("grad_output"), False),
+        )
+
+    @classmethod
+    def _check_fp8_meta(cls, fp8_meta: Optional[dict[str, Any]]) -> None:
+        if fp8_meta is None:
+            return
+
+        # Update FP8 recipe and communication group
+        recipe = FP8GlobalStateManager.get_fp8_recipe()
+        fp8_meta["recipe"] = recipe
+        fp8_meta["fp8_group"] = FP8GlobalStateManager.get_fp8_group()
+
+        # Adjust amax history length if needed
+        amax_history_len = recipe.amax_history_len
+        for is_forward in (True, False):
+            key = FP8GlobalStateManager.get_meta_tensor_key(forward=is_forward)
+            if key not in fp8_meta:
+                continue
+            meta = fp8_meta[key]
+            curr_len = meta.amax_history.size(0)
+            if curr_len == amax_history_len:
+                continue
+            with torch.no_grad():
+                if curr_len > amax_history_len:
+                    meta.amax_history = meta.amax_history[:amax_history_len].clone()
+                else:
+                    meta.amax_history = torch.nn.functional.pad(
+                        meta.amax_history,
+                        pad=(0, 0, 0, amax_history_len - curr_len),
+                    )
+
+    def get_fp8_meta(self, mode: str) -> Optional[dict[str, Any]]:
+        """FP8 metadata
+
+        Parameters
+        ----------
+        mode: {"input", "param", "grad_output"}
+            Type of FP8 scaling factor
+
+        """
+        if self._fp8_metas is None:
+            self._fp8_metas = self._make_fp8_metas()
+        return self._fp8_metas[mode]
+
+    def pre_forward(self) -> None:
+        """Preprocessing before forward pass"""
+
+        # Initialize FP8 metadata if needed
+        fp8_enabled = FP8GlobalStateManager.is_fp8_enabled()
+        if fp8_enabled:
+
+            # Construct FP8 metadata if needed
+            if self._fp8_metas is None:
+                self._fp8_metas = self._make_fp8_metas()
+
+            # Make sure FP8 metadata matches FP8 autocast context
+            # if any(fp8_meta is not None for fp8_meta in self._fp8_metas):
+            for fp8_meta in self._fp8_metas.values():
+                self._check_fp8_meta(fp8_meta)
+
+            # Register FP8 metadata for amax and scale update
```

Comment:
Exactly:
https://github.com/NVIDIA/TransformerEngine/blob/7326af9d8d7f7a9d2a4d24b0193d5bb51541a80d/transformer_engine/pytorch/module/base.py#L608-L611
Although now that you mention it, we should register `"grad_output"` in the backward pass instead of the forward.

Reply:
Actually, the current implementation matches the module API. The `fp8_meta`s are registered in the forward pass, and we manually trigger an update in the backward pass:

https://github.com/timmoon10/TransformerEngine/blob/f4e6af92e8956d948fe1fbaefbc1b2dd6f32b457/transformer_engine/pytorch/ops/fuser.py#L169-L171

https://github.com/NVIDIA/TransformerEngine/blob/7326af9d8d7f7a9d2a4d24b0193d5bb51541a80d/transformer_engine/pytorch/module/linear.py#L612-L613